### PR TITLE
refactor(request-response): don't use upgrade infrastructure

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

This patch refactors `libp2p-request-response` to not use the "upgrade infrastructure" provided by `libp2p-swarm`. Instead, we directly convert the negotiated streams into futures that read and write the messages.

Related: #3268.
Related: #2863.

## Attributions

Co-authored-by: Yiannis Marangos <yiannis@eiger.co>

## Notes & open questions

~I am putting this up as an initial draft. I think some clean-up can be done but overall, I am glad it is working this well.~

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/request-response/src/handler.rs
Comment (mxinden) [inline protocols/request-response/src/handler.rs:143]:
```suggestion
            let request = codec.read_request(&protocol, &mut stream).await?;
```

Wouldn't that compile as well?

File: protocols/request-response/src/handler.rs
Comment (mxinden) [inline protocols/request-response/src/handler.rs]:
```suggestion
                    panic!("Expect request receiver to be alive i.e. connection handler to be alive.",)
```

**Review by mxinden — COMMENTED** [review]
Direction looks good to me! Assuming that you don't want a final review yet.

**Comment by thomaseizinger** [comment]
> Direction looks good to me! Assuming that you don't want a final review yet.

A high-level review would be appreciated :)

**Review by mxinden — COMMENTED** [review]
Direction looks good to me.

File: protocols/request-response/src/handler.rs
Comment (thomaseizinger) [reply protocols/request-response/src/handler.rs:143]: Probably, I just verbatim copied what the previous code did.

File: protocols/request-response/src/handler.rs
Comment (thomaseizinger) [inline protocols/request-response/src/handler.rs]: This is wrong, `multistream-select` can time out.

File: protocols/request-response/src/handler.rs
Comment (thomaseizinger) [reply protocols/request-response/src/handler.rs:143]: Causes some kind of `Send`/`Sync` error with `protocol`.

File: protocols/request-response/src/handler.rs
Comment (thomaseizinger) [inline protocols/request-response/src/handler.rs]: Blocked by https://github.com/libp2p/rust-libp2p/pull/4029.

Comment (thomaseizinger) [reply protocols/request-response/src/handler.rs]: This isn't necessarily blocked actually. Perhaps we can remove some items from this dependency chain here but finding a different solution.

**Comment by thomaseizinger** [comment]
@mxinden This is ready for review and is the 2nd last piece towards starting the work around a simpler design of `ConnectionHandler`s.

File: protocols/request-response/src/handler.rs
Comment (mxinden) [inline protocols/request-response/src/handler.rs]: Is this bounded somehow? Previously we would leverage the `libp2p-swarm` [max_negotiating_inbound_streams](https://docs.rs/libp2p-swarm/latest/libp2p_swarm/struct.SwarmBuilder.html#method.max_negotiating_inbound_streams) for inbound streams.

File: protocols/request-response/src/handler.rs
Comment (mxinden) [inline protocols/request-response/src/handler.rs:65]: Why request more than one new outbound stream at a time? Can any of our muxers even do any work here concurrently. Or does this just make the implementation simpler?

**Review by mxinden — COMMENTED** [review]
We need to bound the number of concurrent inbound streams. Otherwise this change looks good to me.

(Surprised that this ends up being 2 lines less instead of many new lines in addition.)

File: protocols/request-response/src/handler.rs
Comment (thomaseizinger) [reply protocols/request-response/src/handler.rs:65]: I think if you are not using 0RTT stream negotiation, you'll pay a lot of round-trips if you do these streams sequentially because `Connection` won't be giving you a stream until it is fully negotiated.

I think I prefer the implementation this way because it doesn't assume anything about the muxers. The user may request a series of streams in a row so it makes sense to just forward that instead of handling it one by one.

File: protocols/request-response/src/handler.rs
Comment (thomaseizinger) [reply protocols/request-response/src/handler.rs]: At the moment there is no limit. What do you think is an appropriate limit to set here?

Comment (thomaseizinger) [reply protocols/request-response/src/handler.rs]: I've introduced a crate `futures-bounded` that I am planning to use for housing utilities around size and time bounded worker futures.

Comment (thomaseizinger) [reply protocols/request-response/src/handler.rs]: I've chosen an arbitrary limit of 100 for now but `futures_bounded::WorkerFutures` is ready to work with backpressure once we have it. For now, any additional streams will be dropped.

Comment (mxinden) [reply protocols/request-response/src/handler.rs]: `100` as a default sounds good to me.

I am afraid we will have to make this configurable. E.g. Substrate has chosen a very high `max_negotiating_inbound_streams` limit in order for `libp2p-request-response` to not hit any limits.

https://github.com/paritytech/substrate/blob/a6b0d435995d7101e70ee0aab1707d0807024a3f/client/network/src/service.rs#L472

//CC @kpp and @melekes

File: protocols/request-response/src/handler.rs
Comment (mxinden) [reply protocols/request-response/src/handler.rs:65]: Makes sense. Thank you @thomaseizinger for expanding on this.

File: misc/futures-bounded/src/lib.rs
Comment (mxinden) [inline misc/futures-bounded/src/lib.rs]:
```suggestion
/// This wraps [`FuturesUnordered`] but bounds it by time and size.
```

File: misc/futures-bounded/Cargo.toml
Comment (mxinden) [inline misc/futures-bounded/Cargo.toml]: Not really bounding `Future`s in size, but rather the number, no?

**Review by mxinden — COMMENTED** [review]
`futures-bounded` looks good to me. Thank you!

File: misc/futures-bounded/src/lib.rs
Comment (thomaseizinger) [reply misc/futures-bounded/src/lib.rs]: I think this still renders as a link but I'll try!

File: misc/futures-bounded/Cargo.toml
Comment (thomaseizinger) [reply misc/futures-bounded/Cargo.toml]: With size in this case I was referring to memory size, i.e. we are bounding the total member that a set of futures can consume by limiting how many we can have at a time. Has a nice ring to it I find :D

File: protocols/request-response/src/handler.rs
Comment (thomaseizinger) [reply protocols/request-response/src/handler.rs]: Those are for the overall `Swarm` though whereas here we are talking about the limit for a single protocol. We can still make it configurable though, I think adding an option here is not a breaking change. I'll look into it!

Comment (thomaseizinger) [reply protocols/request-response/src/handler.rs]: I think having to configure such a limit per protocol now is a slight decrease in usability. I still think that this is the right way forward though. Hopefully, we can figure out a more generic approach in the future.

File: misc/futures-bounded/src/lib.rs
Comment (thomaseizinger) [reply misc/futures-bounded/src/lib.rs]: Works:

image

Comment (mxinden) [reply misc/futures-bounded/src/lib.rs]: Don't you want `FuturesUnordered` to be monospaced to highlight, that it is a symbol?

File: misc/futures-bounded/Cargo.toml
Comment (mxinden) [reply misc/futures-bounded/Cargo.toml]: The description above does not mention `set` or `group` or `collection`. Thus with `futures` I am confusing the `Future` added to the set.

Either way. Works for me.

File: protocols/request-response/CHANGELOG.md
Comment (mxinden) [inline protocols/request-response/CHANGELOG.md]:
```suggestion
- Allow at most 100 concurrent inbound + outbound request-response protocol streams per connection.
```

**Review by mxinden — COMMENTED** [review]
I am hesitant to publish this as a patch release.

@kpp and @melekes will need to change code in https://github.com/paritytech/substrate/ in order to upgrade to this patch release. In other words Substrate will need to set `Config::with_max_concurrent_streams` in order to maintain semantic behavior, that is to continue supporting many inbound streams at once.

In my eyes the above implies that this is a breaking change.

Thoughts?

**Comment by thomaseizinger** [comment]
> @kpp and @melekes will need to change code in [paritytech/substrate](https://github.com/paritytech/substrate/) in order to upgrade to this patch release. In other words Substrate will need to set `Config::with_max_concurrent_streams` in order to maintain semantic behavior, that is to continue supporting many inbound streams at once.
> 
> In my eyes the above implies that this is a breaking change.
> 
> Thoughts?

This is why modern package managers have lockfiles: To ensure that even small changes don't just randomly break your code.

From an API PoV, this is not a breaking change. I see your point but I disagree. If we can't publish new features as semver-compatible updates, what is the point of making them?

Even if it is "more" correct that this is a breaking change, do you think that it will actually be more visible to users if we batch it together with other breaking changes? The contrary might be the case. A single update of just `libp2p-request-response` might be much easier to isolate in case something actually breaks than figuring out what is causing failures in a `-1000 +1000` update PR to the next minor release of `libp2p`.

Semantic versioning rules are for machines, not humans. We need to make sure that our semver bumps are inline with cargo's dependency resolution algorithm in order to not cause unexpected build failures. What is actually contained in an update and how that affects your application is quite detached from that. A trivial bugfix can "break" your application if you happen to rely on the previous behaviour in some weird way.

> need to set `Config::with_max_concurrent_streams` in order to maintain semantic behavior, that is to continue supporting many inbound streams at once.

We can set it to a higher default value to reduce the risk of this actually breaking someone. A single stream consumes little memory so we could easily set this to 1000 I think.

**Comment by mxinden** [comment]
> Semantic versioning rules are for machines, not humans.

I disagree. It informs a downstream user whether they can upgrade to a new version without changing the behavior of their application.

> Minor version Y (x.Y.z | x > 0) MUST be incremented if new, backward compatible functionality is introduced to the public API.

https://semver.org/#spec-item-7

The semantic functionality change in this pull request is not backwards compatible. In other words it is not safe for a downstream user to upgrade to this version. 

> If we can't publish new features as semver-compatible updates, what is the point of making them?

The introduction of `with_max_concurrent_streams` is a new feature. No longer leveraging the `libp2p-swarm` upgrade mechanism and thus no longer being limited by `max_negotiating_inbound_streams` is a breaking change.

> Even if it is "more" correct that this is a breaking change, do you think that it will actually be more visible to users if we batch it together with other breaking changes?

Yes. I think it will be more visible. Downstream users do not need to read changelogs of non-breaking releases. The opposite is the case for breaking releases.


> > need to set `Config::with_max_concurrent_streams` in order to maintain semantic behavior, that is to continue supporting many inbound streams at once.
> 
> We can set it to a higher default value to reduce the risk of this actually breaking someone. A single stream consumes little memory so we could easily set this to 1000 I think.

We can not. We allocate more than `256KiB` (Yamux default receive window + various other buffers) per stream. I don't think `1000` streams times `256 KiB` times e.g. `100` connections  totaling at `25.6 GiB` is a DOS safe default.

That is why we chose a small `max_negotiating_inbound_streams`:

https://github.com/libp2p/rust-libp2p/blob/8937105e07647dbd351f7a66601083e81fb10b22/swarm/src/connection/pool.rs#L967

Note that I don't agree with Substrate's default value, i.e. I don't think it is a good default. That said, I don't think that warrants breaking it.

I know this makes the larger refactoring even more difficult and time consuming.

**Comment by thomaseizinger** [comment]
> > Minor version Y (x.Y.z | x > 0) MUST be incremented if new, backward compatible functionality is introduced to the public API.
> 
> [semver.org/#spec-item-7](https://semver.org/#spec-item-7)

FWIW, we are `x < 0` so technically, this does not apply to us.

> > Even if it is "more" correct that this is a breaking change, do you think that it will actually be more visible to users if we batch it together with other breaking changes?
> 
> Yes. I think it will be more visible. Downstream users do not need to read changelogs of non-breaking releases. The opposite is the case for breaking releases.

I am not sure how much we can rely on this in practice. In my experience, most Rust users will compile + run their tests on version upgrades (that is what we do too) and if everything passes, the version upgrade is merged.

Personally, I'd only go and read changelogs if something breaks and I need more information. Would you bother and read a changelog entry for a minor version bump if everything compiles just fine? I don't think I would.

If we want to make this a breaking change, I think we need to think of a way of actually breaking compilation. I suspect that most users won't look at it otherwise.

**Comment by thomaseizinger** [comment]
> `futures-bounded` looks good to me. Thank you!

Would you agree to merge this in a separate PR to not block other work like https://github.com/libp2p/rust-libp2p/pull/4275 on this PR?

**Comment by mxinden** [comment]
> > `futures-bounded` looks good to me. Thank you!
> 
> Would you agree to merge this in a separate PR to not block other work like #4275 on this PR?

Fine by me :+1:

**Comment by thomaseizinger** [comment]
@mxinden Following our sync conversation, you want to publish this as a breaking change but be API compatible and ping individual users - most notable substrate - directly when they perform they upgrade to set this limit correctly, yes?

We could do the same with a patch-release. It might not be semver-correct by some interpretation (the official spec says "anything can change at any time" before 1.0).

I do think that it might be unexpected by some users so I am happy to delay the change to the next minor release.

File: misc/futures-bounded/src/lib.rs
Comment (dgarus) [inline misc/futures-bounded/src/lib.rs]: Hi, @thomaseizinger !

Why do we need the key parameter here?
When I see the key, I think that I can get or cancel something (future, for example) by this key.
But not in this case.
I understand that the idea is distinguishing results, so maybe to call this parameter `worker_id`?

Comment (thomaseizinger) [reply misc/futures-bounded/src/lib.rs]: Sure, I think the local variable name doesn't matter much :)

**Comment by mxinden** [comment]
> > > `futures-bounded` looks good to me. Thank you!
> > 
> > 
> > Would you agree to merge this in a separate PR to not block other work like #4275 on this PR?
> 
> Fine by me 👍

Contrary to what we have discussed out-of-band @thomaseizinger I have not merged `futures-bounded` into `master` in a separate pull request. I think it makes more sense to merge it along with either this pull request or https://github.com/libp2p/rust-libp2p/pull/4275/. That way we know it's API is sufficient. If I merge it now already, it might be unnecessarily released. As far as I know no one else is blocked on it being available.

**Comment by thomaseizinger** [comment]
> > > > `futures-bounded` looks good to me. Thank you!
> > > 
> > > Would you agree to merge this in a separate PR to not block other work like #4275 on this PR?
> > 
> > Fine by me 👍
> 
> Contrary to what we have discussed out-of-band @thomaseizinger I have not merged `futures-bounded` into `master` in a separate pull request. I think it makes more sense to merge it along with either this pull request or #4275. That way we know it's API is sufficient. If I merge it now already, it might be unnecessarily released. As far as I know no one else is blocked on it being available.

Yeah that is fine now that we have copied it into #4275.

File: protocols/request-response/CHANGELOG.md
Comment (thomaseizinger) [reply protocols/request-response/CHANGELOG.md]: This isn't quite true because the limit applies per _instance_ of `request-response`.

**Comment by thomaseizinger** [comment]
Will need to be updated once https://github.com/libp2p/rust-libp2p/pull/4275 is merged to use the `futures-bounded` version that is embedded there now.

File: protocols/request-response/src/lib.rs
Comment (thomaseizinger) [inline protocols/request-response/src/lib.rs:310]: @mxinden Would you agree that we can publish this as a patch release if we link to an issue / migration document here that explains that we are now capping streams at 100?

File: swarm/CHANGELOG.md
Comment (thomaseizinger) [inline swarm/CHANGELOG.md]:
```suggestion

```

File: protocols/request-response/src/handler.rs
Comment (thomaseizinger) [inline protocols/request-response/src/handler.rs:243]: Not sure I still agree with this behaviour actually. I think @oblique wanted to have this reported as an error. That is probably a better idea than naively retrying!

Comment (oblique) [reply protocols/request-response/src/handler.rs:243]: @thomaseizinger Yes I wanted this reported as an error because `Codec` can fail on deserialization of the response.

**Comment by thomaseizinger** [comment]
This needs some more thinking on how we want to report errors. Also, I noticed that several of the enum variants are actually unused so we should properly clean this up.

I'll see to look into it next week. Help much appreciated though :)

**Comment by mxinden** [comment]
@thomaseizinger let me know once this is ready for another review.

**Comment by thomaseizinger** [comment]
> @thomaseizinger let me know once this is ready for another review.

It is! :)

**Review by mxinden — APPROVED** [review]
Great work! Happy for this to land.

**Pull request merged**
